// mcsmanager-common/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceStreamListener = exports.GlobalVariable = exports.StorageSubsystem = exports.MCServerStatus = exports.removeTrail = exports.arrayUnique = exports.supposeValue = exports.isEmpty = exports.toNumber = exports.toBoolean = exports.toText = exports.configureEntityParams = exports.QueryWrapper = exports.LocalFileSource = exports.MySqlSource = exports.QueryMapWrapper = exports.systemInfo = exports.killProcess = exports.ProcessWrapper = void 0;
const system_storage_1 = __importDefault(require("./system_storage"));
exports.StorageSubsystem = system_storage_1.default;
const global_variable_1 = __importDefault(require("./global_variable"));
exports.GlobalVariable = global_variable_1.default;
const instance_stream_1 = __importDefault(require("./instance_stream"));
exports.InstanceStreamListener = instance_stream_1.default;
const mcping_1 = __importDefault(require("./mcping"));
exports.MCServerStatus = mcping_1.default;
var process_tools_1 = require("./process_tools");
Object.defineProperty(exports, "ProcessWrapper", { enumerable: true, get: function () { return process_tools_1.ProcessWrapper; } });
Object.defineProperty(exports, "killProcess", { enumerable: true, get: function () { return process_tools_1.killProcess; } });
var system_info_1 = require("./system_info");
Object.defineProperty(exports, "systemInfo", { enumerable: true, get: function () { return system_info_1.systemInfo; } });
var query_wrapper_1 = require("./query_wrapper");
Object.defineProperty(exports, "QueryMapWrapper", { enumerable: true, get: function () { return query_wrapper_1.QueryMapWrapper; } });
Object.defineProperty(exports, "MySqlSource", { enumerable: true, get: function () { return query_wrapper_1.MySqlSource; } });
Object.defineProperty(exports, "LocalFileSource", { enumerable: true, get: function () { return query_wrapper_1.LocalFileSource; } });
Object.defineProperty(exports, "QueryWrapper", { enumerable: true, get: function () { return query_wrapper_1.QueryWrapper; } });
var typecheck_1 = require("./typecheck");
Object.defineProperty(exports, "configureEntityParams", { enumerable: true, get: function () { return typecheck_1.configureEntityParams; } });
Object.defineProperty(exports, "toText", { enumerable: true, get: function () { return typecheck_1.toText; } });
Object.defineProperty(exports, "toBoolean", { enumerable: true, get: function () { return typecheck_1.toBoolean; } });
Object.defineProperty(exports, "toNumber", { enumerable: true, get: function () { return typecheck_1.toNumber; } });
Object.defineProperty(exports, "isEmpty", { enumerable: true, get: function () { return typecheck_1.isEmpty; } });
Object.defineProperty(exports, "supposeValue", { enumerable: true, get: function () { return typecheck_1.supposeValue; } });
var array_1 = require("./array");
Object.defineProperty(exports, "arrayUnique", { enumerable: true, get: function () { return array_1.arrayUnique; } });
var string_utils_1 = require("./string_utils");
Object.defineProperty(exports, "removeTrail", { enumerable: true, get: function () { return string_utils_1.removeTrail; } });
